source/common/network/lc_trie.h

```diff
@@ -0,0 +1,487 @@
+     * @param first supplies the index into the ip_prefixes_ vector where to start building the
+     * trie_ from.
+     * @param n suppplies the number of entries.
+     * @param pos supplies the root for this trie..
+     * @param nextfree supplies the next available index in the trie_.
+     */
+    void buildRecursive(uint32_t prefix, uint32_t first, uint32_t n, uint32_t pos,
+                        uint32_t* nextfree) {
+      uint32_t p, address, bits;
+      uint32_t bit_pattern;
+      int k;
+
+      // Setting a leaf, the branch and skip are 0.
+      if (n == 1) {
+        trie_[pos] = first;
+        return;
+      }
+
+      ComputePair output = computeBranch(prefix, first, n);
+
+      address = *nextfree;
+      trie_[pos] =
+          setBranch(output.branch_) | setSkip(output.prefix_ - prefix) | setAddress(address);
+
+      // For the next part of the trie, the next available address is at address + 2^branching
+      // factor.
+      *nextfree += 1 << output.branch_;
+      p = first;
+
+      // Build the subtrees.
+      for (bit_pattern = 0; bit_pattern < static_cast<uint32_t>(1 << output.branch_);
+           ++bit_pattern) {
+
+        k = 0;
+        // k is finding the number of entries in the base vector that have the same bit pattern.
+        while (p + k < first + n &&
+               static_cast<uint32_t>(extractBits<IpType, address_size>(
+                   output.prefix_, output.branch_, ip_prefixes_[p + k].ip_)) == bit_pattern) {
+          ++k;
+        }
+
+        if (k == 0) {
+          if (p == first + n) {
+            buildRecursive(output.prefix_ + output.branch_, p - 1, 1, address + bit_pattern,
+                           nextfree);
+          } else {
+            buildRecursive(output.prefix_ + output.branch_, p, 1, address + bit_pattern, nextfree);
+          }
+        } else if (k == 1 && ip_prefixes_[p].length_ - output.prefix_ < output.branch_) {
+          uint32_t i;
+          bits = output.branch_ + output.prefix_ - ip_prefixes_[p].length_;
+          for (i = bit_pattern; i < bit_pattern + (1 << bits); ++i) {
+            buildRecursive(output.prefix_ + output.branch_, p, 1, address + i, nextfree);
+          }
+        } else {
+          buildRecursive(output.prefix_ + output.branch_, p, k, address + bit_pattern, nextfree);
+        }
+        p += k;
+      }
+    }
+
+    /**
+     * LcNode is a uint32_t. A wrapper is provided to simplify getting the branch, skip and
+     * address values held within the data structure.
+     *
+     * The LcNode has three parts to it
+     * - Branch: the first 5 bits represent the branching factor. This supports a branching factor
+     * of 2^31.
+     * - Skip: the next 7 bits represent the number of bits to skip when looking at an IP address.
+     * This value can be between 0 and 127, so IPv6 is supported as well.
+     * - Address: the remaining 30 bits represent the pointer. If branch value is not zero, this
```

Comment:
If I'm counting right, it seems like there are only 20 remaining bits.  32 bits minus 5 for the branching factor, minus 7 for the skip count.

Reply:
Yes, that is a typo. I'll fix that.